pkg/controller/anomalydetector/controller.go

```diff
@@ -0,0 +1,632 @@
+// Copyright 2022 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package anomalydetector
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"reflect"
+	"regexp"
+	"sync"
+	"time"
+
+	apimachineryerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	apimachinerytypes "k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	crdv1alpha1 "antrea.io/theia/pkg/apis/crd/v1alpha1"
+	"antrea.io/theia/pkg/client/clientset/versioned"
+	crdv1a1informers "antrea.io/theia/pkg/client/informers/externalversions/crd/v1alpha1"
+	"antrea.io/theia/pkg/client/listers/crd/v1alpha1"
+	"antrea.io/theia/pkg/util"
+	"antrea.io/theia/pkg/util/clickhouse"
+	sparkv1 "antrea.io/theia/third_party/sparkoperator/v1beta2"
+)
+
+const (
+	controllerName = "AnomalyDetectorController"
+	// Set resyncPeriod to 0 to disable resyncing.
+	resyncPeriod time.Duration = 0
+	// How long to wait before retrying the processing of an Service change.
+	minRetryDelay = 5 * time.Second
+	maxRetryDelay = 300 * time.Second
+	// Default number of workers processing an Service change.
+	defaultWorkers = 4
+	// Time format for parsing input time
+	inputTimeFormat  = "2006-01-02 15:04:05"
+	k8sQuantitiesReg = "^([+-]?[0-9.]+)([eEinumkKMGTP]*[-+]?[0-9]*)$"
+	// Spark related parameters
+	sparkImage           = "projects.registry.vmware.com/antrea/theia-anomaly-detection:latest"
+	sparkImagePullPolicy = "IfNotPresent"
+	sparkAppFile         = "local:///opt/spark/work-dir/AnomalyDetection.py"
+	sparkServiceAccount  = "theia-spark"
+	sparkVersion         = "3.1.1"
+	sparkPort            = 4040
```

Comment:
Maybe we should move these variables (except sparkAppFile) to util package so that if there are some changes, we do not need to update them at both tad controller and npr controller

Reply:
> Maybe we should move these variables (except sparkAppFile) to util package so that if there are some changes, we do not need to update them at both tad controller and npr controller

Ditto. Highlight in case you miss it.

Reply:
Sure, this is done as a part of #161  as the spark dockerfiles and its related variables are altered in that commit. I shall be uploading the commit for review soon for that